[WebGPU] de-couple vertex buffer binding indices from bind group binding indices (251185)

mwyrzykowski opened the pull request:
#### 1303b0d0d672202dce205e4d17996955ba96be3d
<pre>
[WebGPU] de-couple vertex buffer binding indices from bind group binding indices (251185)
<a href="https://bugs.webkit.org/show_bug.cgi?id=251185">https://bugs.webkit.org/show_bug.cgi?id=251185</a>
&lt;radar://104673394&gt;

Reviewed by Myles C. Maxfield.

The shader compiler needs to know what offset to start generating
buffer indices from, so partition the available buffers into two ranges:
one for stage_in or calls to setVertexBuffer and the other for uniform bind groups.

* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::setBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setBindGroup):
(WebGPU::RenderPassEncoder::setPipeline):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::device const):
(WebGPU::RenderPipeline::vertexShaderInputBufferCount const): Deleted.
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js:
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-indirect.js:
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/hello-cube.js:
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/hello-indexed-cube.js:
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/indirect-command-buffer-textured-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/query-sets.js:
(async helloTriangle.async frameUpdate):
(async helloTriangle):
* Websites/webkit.org/demos/webgpu/scripts/textured-cube-vs-input-buffers.js:
* Websites/webkit.org/demos/webgpu/scripts/two-cubes.js:
(async helloCube.frameUpdate):
(async helloCube):

Canonical link: <a href="https://commits.webkit.org/259414@main">https://commits.webkit.org/259414@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb0baa2d06ead17327a7e582e86a33c8f720c9df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104853 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13933 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37749 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114122 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15062 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4857 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97180 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113151 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110613 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94645 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/39159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93502 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26261 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7277 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27622 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7374 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-WK2-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13429 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47172 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6493 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9163 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/RenderPassEncoder.mm:
```diff
@@ -177,7 +177,7 @@
         [m_renderCommandEncoder useResource:resource.mtlResource usage:resource.usage stages:resource.renderStages];
 
-    [m_renderCommandEncoder setVertexBuffer:group.vertexArgumentBuffer() offset:0 atIndex:groupIndex + m_vertexShaderInputBufferCount];
+    [m_renderCommandEncoder setVertexBuffer:group.vertexArgumentBuffer() offset:0 atIndex:groupIndex + m_device->limits().maxVertexBuffers];
```
I thought we were going to have this count down from the max?

mwyrzykowski replied:
oh like `m_device->limits().maxVertexBuffers - groupIndex - 1` right?

litherum replied:
Right, for now. When we get the real limit we'll adjust accordingly.

litherum reviewed (approved):
Thanks for fixing up all the examples

mwyrzykowski committed:
[WebGPU] de-couple vertex buffer binding indices from bind group binding indices (251185)
https://bugs.webkit.org/show_bug.cgi?id=251185
<radar://104673394>

Reviewed by Myles C. Maxfield.

The shader compiler needs to know what offset to start generating
buffer indices from, so partition the available buffers into two ranges:
one for stage_in or calls to setVertexBuffer and the other for uniform bind groups.

* Source/WebGPU/WebGPU/RenderBundleEncoder.mm:
(WebGPU::RenderBundleEncoder::setBindGroup):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::setBindGroup):
(WebGPU::RenderPassEncoder::setPipeline):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
(WebGPU::RenderPipeline::device const):
(WebGPU::RenderPipeline::vertexShaderInputBufferCount const): Deleted.
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::RenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js:
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-indirect.js:
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/hello-cube.js:
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/hello-indexed-cube.js:
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/indirect-command-buffer-textured-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:
(async helloCube):
* Websites/webkit.org/demos/webgpu/scripts/query-sets.js:
(async helloTriangle.async frameUpdate):
(async helloTriangle):
* Websites/webkit.org/demos/webgpu/scripts/textured-cube-vs-input-buffers.js:
* Websites/webkit.org/demos/webgpu/scripts/two-cubes.js:
(async helloCube.frameUpdate):
(async helloCube):

Canonical link: https://commits.webkit.org/259414@main
